jupyter_server/gateway/managers.py

```diff
@@ -200,9 +200,28 @@ def _get_endpoint_for_user_filter(default_endpoint):
             return "?user=".join([default_endpoint, kernel_user])
         return default_endpoint
 
+    def _replace_path_kernelspec_resources(self, kernel_specs):
+        """Helper method that replaces any gateway base_url with the server's base_url
+        This enables clients to properly route through jupyter_server to a gateway
+        for kernel resources such as logo files
+        """
+        kernelspecs = kernel_specs["kernelspecs"]
+        for kernel_name in kernelspecs:
+            resources = kernelspecs[kernel_name]["resources"]
+            for resource_name in resources:
+                original_path = resources[resource_name]
+                split_eg_base_url = str.rsplit(original_path, sep="/kernelspecs/", maxsplit=1)
+                kernel_specs["kernelspecs"][kernel_name]["resources"][
+                    resource_name
+                ] = url_path_join(self.parent.base_url, "/kernelspecs", split_eg_base_url[1])
+                self.log.debug(
+                    f"Replaced original kernel resource path {original_path} with new "
+                    f"path {kernel_specs['kernelspecs'][kernel_name]['resources'][resource_name]}"
```

Comment:
I'm noticing the handling of the leading slash is inconsistent.  This doesn't appear to cause issues but could trigger a "wrong path" when folks need to troubleshoot.

For example, for the case when only the server is configured with a `base_url`, this is produced:
```
Replaced original kernel resource path /kernelspecs/spark_scala_kubernetes/logo-64x64.png with new path /jp/foo/kernelspecs/spark_scala_kubernetes/logo-64x64.png
```
whereas when both gateway and server are configured with `base_url`, it yields:
```
Replaced original kernel resource path eg/bar/kernelspecs/scala_k8s_spark/logo-64x64.png with new path /jp/foo/kernelspecs/scala_k8s_spark/logo-64x64.png
```
and when only gateway is configured with `base_url`...
```
Replaced original kernel resource path eg/bar/kernelspecs/scala_k8s_spark/logo-64x64.png with new path /kernelspecs/scala_k8s_spark/logo-64x64.png
```
Since the entries themselves only start with a `/` in the default case (when `base_url` is not configured), it seems like the `/` is considered part of `/kernelspecs`.  As a result, I don't think it should be included in the logging output when a `base_url` is configured since the resource entries themselves do not include it.

Reply:
I see what you're getting at and agree that it is a bit confusing as it exists now. IMO that is more of a problem with gateway's handling of `base_url`. When one is configured with a leading `/` such as `--EnterpriseGatewayApp.base_url=/sam/base`, the resource paths returned do start with that leading `/`. 
<img width="1392" alt="Screen Shot 2022-12-12 at 3 47 48 PM" src="https://user-images.githubusercontent.com/20909250/207162042-3017bb78-18ef-490f-8795-3f0eaca054e7.png">

By default, when neither servers have `base_url`s, the resource paths also start with a leading `/`. Treating that as the _intended_ path structure, gateway is at fault for not including a leading `/` when it is configured with a `base_url` but the config value doesn't start with that leading `/`, such as `--EnterpriseGatewayApp.base_url=sam/base`. 
<img width="1392" alt="Screen Shot 2022-12-12 at 3 47 43 PM" src="https://user-images.githubusercontent.com/20909250/207162587-d82e5d24-5c63-43bd-b13f-a71af02c04e8.png">

In the three other scenarios, resource entities **do** include the leading `/`, and they will always include it after the path transformation done in this PR due to the behavior of the `url_path_join` util method so I hesitate to modify the log message to strip anything out from the actual value that is being returned.

To your point though, it doesn't appear to impact behavior at all because of how the url paths are put together before calling out to a gateway.

Reply:
Great stuff - thanks for digging into this!  Although EG inherits the handlers, we cache the kernelspecs, so we (EG) should look into how it's handling leading slashes  - I'll open an issue.